### PR TITLE
Allow using password for workbook protection.

### PR DESCRIPTION
**Pull request by nomomax**
#### What's this PR do?
Provide more methods to easily protect/unprotect workbook using password.
#### Where should the reviewer start?
Most work is in `ClosedXML/Excel/XLWorkbook.cs`, `ClosedXML/Excel/XLWorkbook_Load.cs` and `ClosedXML/Excel/XLWorkbook_Save.cs`. 
#### How should this be manually tested?
Load up an excel and use `Workbook.Protect` or `Workbook.Unprotect` and then save and verify the protection of this workbook.
#### Any background context you want to provide?
1. The method `GetPasswordHash` in `ClosedXML/Excel/XLWorkbook.cs` is copied from `ClosedXML/Excel/XLSheetProtection.cs`.

2. The `LockWindows` option is unavailable in Excel 2013 and later versions. https://support.office.com/en-us/article/Protect-a-workbook-7e365a4d-3e89-4616-84ca-1931257c1517

#### Questions:
- Is there a blog post?
- Does the knowledge base need an update?
- Does this add new (C#) dependencies?

- [ ] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

### CODE REVIEW

File: ClosedXML/Excel/XLWorkbook.cs [hunk 1]
```diff
@@ -864,21 +864,96 @@ public static Object EvaluateExpr(String expression)
...
+                    if (lockStructure || lockWindows)
+                    {
+                        //Protect workbook using password.
+                        LockPassword = hashPassword;
+                    }
+                }
+            }
+            else
+            {
+                if (LockPassword != null)
```
Comment (Pyropace) [inline ClosedXML/Excel/XLWorkbook.cs]: ❓ 🔴 you already checking for LockPassword is not above, why are you checking it here again ?

Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: Lower case variable `lockPassword` is different from the upper case variable `LockPassword`.

Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: The checking for 'LockPassword' above is in the case if 'lockPassword' is not null. The checking here is in the case if 'lockPassword' is null.

Comment (Pyropace) [inline ClosedXML/Excel/XLWorkbook.cs]: ⚪️ ah i see, i reviewed it late last night, could you change this and use a different variable name, so its not confusing

**Comment by vbjay** [comment]
I agree. workbookPassword sheetPassword.

File: ClosedXML/Excel/XLWorkbook.cs (hunk 1 above)
Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: Not a problem. @Pyropace @vbjay requested changes have been made.

**Comment by nomomax** [comment]
@Pyropace do i need to resolve these conflicts?

**Comment by Pyropace** [comment]
yes please, keep your branch up to date untill its merged

File: .gitignore [hunk 2]
```diff
@@ -28,3 +28,4 @@ obj/
 _ReSharper*/
 [Tt]est[Rr]esult*
 /packages
+/.vs
```
Comment (igitur) [inline .gitignore]: This not related to this PR. Please revert this change.

Comment (igitur) [inline .gitignore]: This not related to this PR. Please revert this change.

File: ClosedXML/Excel/XLWorkbook.cs [hunk 3]
```diff
@@ -1,884 +1,959 @@
...
+            _loadSource = XLLoadSource.Stream;
+            _originalStream = stream;
+            Load(stream);
+        }
+
+#endregion
+
+#region Nested type: UnsupportedSheet
+
+        internal
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Move all parameter validation to the top of the method.

Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Where did you get the hashing algorithm? Maybe add a comment with the link. I also think this should rather be in `XLHelper`.

File: ClosedXML_Examples/Misc/WorkbookProtection.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using ClosedXML.Excel;
+
+namespace ClosedXML_Examples.Misc
+{
+    public class WorkbookProtection : IXLExample
+    {
+        #region Variables
```
Comment (igitur) [inline ClosedXML_Examples/Misc/WorkbookProtection.cs]: Remove the empty regions, please. I know many examples have them. No idea why... it's just fluff.

**Comment by nomomax** [comment]
@igitur did you comment on this PR? i can not see your comment.

**Comment by igitur** [comment]
Hey @nomomax . I'm reviewing this in detail only now. Please pull in the latest `develop` branch and have a look at my review comments.

File: ClosedXML/Excel/XLWorkbook.cs (hunk 3 above)
Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: ah..actually i mention it above in the section `Any background context you want to provide?`
The method `GetPasswordHash` is copied from `ClosedXML/Excel/XLSheetProtection.cs`. do i need to move them both to `XLHelper`?

Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Oh I see. I think best is to create an internal extension method in `StringExtensions`:

```c#
internal static String HashPassword(this String password)
{
    // ... the rest
}
```

Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: PS: I found another, similar, but slightly different implementation at https://social.msdn.microsoft.com/Forums/sqlserver/en-US/bcdffe05-aa40-45e7-842c-42bc7bd0d4af/how-to-lock-specific-cells-programmatically-while-excel-export?forum=oxmlsdk - will check later whether it's actually the same thing or not.

File: .gitignore (hunk 2 above)
Comment (nomomax) [inline .gitignore]: OK, it's done.

File: ClosedXML/Excel/XLWorkbook.cs (hunk 3 above)
Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: OK, it's done.

File: ClosedXML_Examples/Misc/WorkbookProtection.cs (hunk 4 above)
Comment (nomomax) [inline ClosedXML_Examples/Misc/WorkbookProtection.cs]: OK, it's done.

File: ClosedXML/Excel/XLWorkbook.cs (hunk 3 above)
Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: @igitur All requested changes, except the hashing algorithm, have been made.

Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: To create an internal extension method in `StringExtensions` is indeed a proper workaround because there's nothing to do with EXCEL.

**Comment by igitur** [comment]
The diffs look a lot better now. But I see the problem still exists for `.csproj` files. I'll submit a PR to your branch for them too.

**Comment by nomomax** [comment]
OK, Thanks a lot.

**Comment by igitur** [comment]
@nomomax Please email me directly. I want to take this offline. My username is the same as my Github username. @ gmail

**Comment by igitur** [comment]
Ah at last the diff show correctly. I think the problem was on my side. I use a combination of cygwin and TortoiseGit on a Windows machine. That stuffs up the line endings. 

Please complete the other requested changes and we'll merge soon after that.

File: ClosedXML/Excel/XLWorkbook.cs [hunk 5]
```diff
@@ -864,21 +864,96 @@ public static Object EvaluateExpr(String expression)
         public XLWorkbook SetLockStructure(Boolean value) { LockStructure = value; return this; }
         public Boolean LockWindows { get; set; }
         public XLWorkbook SetLockWindows(Boolean value) { LockWindows = value; return this; }
-
+        internal HexBinaryValue LockPassword { get; set; }
+        
+        public void Protect(Boolean lockStructure, Boolean lockWindows, String workbookPassword)
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Please move parameter validation to the top of the method.

File: ClosedXML/Excel/XLWorkbook.cs [hunk 6]
```diff
@@ -864,21 +864,96 @@ public static Object EvaluateExpr(String expression)
...
+        {
+            Protect(false, false);
+        }
+
+        public void Unprotect(string workbookPassword)
+        {
+            Protect(false, false, workbookPassword);
+        }
+
+        private String GetPasswordHash(String password)
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Move to `StringExtensions` as previously mentioned.

File: ClosedXML/Excel/XLWorkbook.cs (hunk 5 above)
Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: Sorry, i don't get it. Could you please give me an example for this?

Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Ok, I'll fix it.

File: ClosedXML/Excel/XLWorkbook.cs [hunk 7]
```diff
@@ -864,21 +864,96 @@ public static Object EvaluateExpr(String expression)
...
+                    }
+                    else
+                    {
+                        if (lockStructure || lockWindows)
+                        {
+                            throw new InvalidOperationException("The workbook is already protected");
+                        }
+                        else
+                        {
+                            //Unprotect workbook using password.
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Why are we UNprotecting here?

Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: We are UNprotecting here because `Unprotect` here means to have parameters `lockStructure` and `lockWindows` both FALSE.

Comment (nomomax) [inline ClosedXML/Excel/XLWorkbook.cs]: Workbook still be in protection while saving if we don't set variable `LockPassword` to be null.

File: ClosedXML/Extensions.cs [hunk 8]
```diff
@@ -132,6 +132,22 @@ public static String ToProper(this String value)
 
             return value.Substring(0, 1).ToUpper() + value.Substring(1);
         }
+
+        public static String GetPasswordHash(String password)
```
Comment (igitur) [inline ClosedXML/Extensions.cs]: Use an extension method, not a static method. Actually, undo this, and merge https://github.com/nomomax/ClosedXML/pull/4

Comment (nomomax) [inline ClosedXML/Extensions.cs]: OK, it's done.

**Comment by nomomax** [comment]
@igitur Finally...all requested changes have been completed.

**Comment by igitur** [comment]
Merged. Thanks for your contribution.

**Comment by nomomax** [comment]
Thank you as well.
